cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java

```diff
@@ -93,12 +95,21 @@ public ClassLoader createProgramClassLoader(CConfiguration cConf, ProgramType ty
       cConf.getBoolean(SparkRuntimeUtils.SPARK_STREAMING_CHECKPOINT_REWRITE_ENABLED);
     boolean rewriteYarnClient = cConf.getBoolean(Constants.AppFabric.SPARK_YARN_CLIENT_REWRITE);
 
+    ClassLoader sparkClassLoader = null;
     try {
-      ClassLoader sparkClassloader = sparkClassloader = createClassLoader(filterScalaClasses, rewriteYarnClient,
-                                                                          rewriteCheckpointTempFileName);
-      return new FilterClassLoader(sparkClassloader, SparkResourceFilters.SPARK_PROGRAM_CLASS_LOADER_FILTER);
+      sparkClassLoader = createClassLoader(filterScalaClasses, rewriteYarnClient, rewriteCheckpointTempFileName);
     } catch (IOException e) {
-      throw Throwables.propagate(e);
+      throw new RuntimeException(e);
+    }
+    // SparkResourceFilter must be instantiated using the above classloader as it has the
+    // org.apache.spark.streaming.StreamingContext class, otherwise it will cause NoClassDefFoundError at runtime
+    // because the parent classloader does not have Spark resources loaded.
+    try {
+      FilterClassLoader.Filter sparkClassLoaderFilter = (FilterClassLoader.Filter) sparkClassLoader
+        .loadClass(SparkResourceFilter.class.getName()).newInstance();
+      return new FilterClassLoader(sparkClassLoader, sparkClassLoaderFilter);
+    } catch (Exception e) {
```

Comment:
Can you combine this try catch block with the previous one? The `loadClass` and `newInstance` methods throw specific exception that you can catch them to distinguish the cases.